kirc.c

```diff
@@ -916,36 +998,113 @@ static void handle_user_input(state l)
     }
     printf("\r\x1b[0K");
     switch (l->buf[0]) {
-    case '/':                  /* send system command */
-        if (l->buf[1] == '#') {
+    case '/':			/* send system command */
+	if(!strncmp(l->buf + 1, "JOIN", 4)||!strncmp(l->buf + 1, "join", 4)){
+	    if(!strchr(l->buf, '#')){
+		printf("\x1b[35m%s\x1b[0m\r\n", l->buf);
+		printf("\x1b[35mIllegal channel!\x1b[0m\r\n");
+		return;
+	    }
+	    chan = strchr(l->buf, '#');
+	    chan ++;
+	    strcpy(l->prompt, chan);
+	    raw("join #%s\r\n", chan);
+	    printf("\x1b[35m%s\x1b[0m\r\n", l->buf);
+	    printf("\x1b[35mJoined #%s!\x1b[0m\r\n", chan);
+	    return;
+	}
+	if(!strncmp(l->buf + 1, "PART", 4)||!strncmp(l->buf + 1, "part", 4)){
+	    tok = strchr(l->buf, '#');
+	    if(strlen(l->buf) == 5){
+        	raw("part #%s\r\n", chan);
+		printf("\x1b[35m%s\x1b[0m\r\n", l->buf);
+        	printf("\x1b[35mLeft #%s!\r\n", chan);
+            	printf("\x1b[35mYou need to use /join or /# to speak in a channel!\x1b[0m\r\n");
+            	chan = NULL;
+            	strcpy(l->prompt, "");
+		return;
+	    }
+	    if(tok){
+                raw("part %s\r\n", tok);
+                printf("\x1b[35m%s\x1b[0m\r\n", l->buf);
+                printf("\x1b[35mLeft %s!\r\n", tok);
+                printf("\x1b[35mYou need to use /join or /# to speak in a channel!\x1b[0m\r\n");
+                chan = NULL;
+                strcpy(l->prompt, "");
+	        return;
+	    }
+	    int ok = 1;
+	    tok = l->buf + 5;
+	    while(*tok){
+	        if(*tok!=' '){
+		    ok = 0;
+		    break;
+	        }
+	        tok ++;
+	    }
+	    if(ok){
+	        chan = l->prompt;
+                raw("part #%s\r\n", chan);
+                printf("\x1b[35m%s\x1b[0m\r\n", l->buf);
+                printf("\x1b[35mLeft #%s!\r\n", chan);
+                printf("\x1b[35mYou need to use /join or /# to speak in a channel!\x1b[0m\r\n");
+                chan = NULL;
+                strcpy(l->prompt, "");
+	        return;
+	    }
+	    printf("\x1b[35m%s\x1b[0m\r\n", l->buf);
+	    printf("\x1b[35mIllegal channel!\x1b[0m\r\n");
+	    return;
+	}
+    	if(l->buf[1]=='/'){
+	    raw("privmsg #%s :%s\r\n", l->prompt, l->buf + 3);
+	    printf("\x1b[35mprivmsg #%s :%s\x1b[0m\r\n", l->prompt, l->buf + 3);
+	    return;
+	}
+	if(!strncmp(l->buf+1, "MSG", 3)||!strncmp(l->buf+1, "msg", 3)){
+	    strtok_r(l->buf + 5, " ", &tok);
+	    if(*(tok+strlen(tok)+1))
+		*(tok+strlen(tok))=' ';
+	    raw("privmsg %s :%s\r\n", l->buf + 5, tok);
+	    if(strncmp(l->buf + 5, "NickServ", 8))
```

Comment:
why is `NickServ` special-cased here?

Reply:
Because you usually send out passwords to NickServ, and you don't want to print the password back to the screen in case someone sees it. Also NickServ responds with good messages so you don't need confirmation of sending something.